src/compiler/utilities.ts

```diff
@@ -10761,3 +10777,577 @@ export function createEvaluator({ evaluateElementAccessExpression, evaluateEntit
     }
     return evaluate;
 }
+
+/** @internal */
+export function isConstAssertion(location: Node) {
+    return (isAssertionExpression(location) && isConstTypeReference(location.type))
+        || (isJSDocTypeTag(location) && isConstTypeReference(location.typeExpression));
+}
+
+/** @internal */
+export function findConstructorDeclaration(node: ClassLikeDeclaration): ConstructorDeclaration | undefined {
+    const members = node.members;
+    for (const member of members) {
+        if (member.kind === SyntaxKind.Constructor && nodeIsPresent((member as ConstructorDeclaration).body)) {
+            return member as ConstructorDeclaration;
+        }
+    }
+}
+
+/** @internal */
+export function createNameResolver({
+    compilerOptions,
+    requireSymbol,
+    argumentsSymbol,
+    error,
+    getSymbolOfDeclaration,
+    globals,
+    lookup,
+    getNodeLinks,
+    onPropertyWithInvalidInitializer,
+    onFailedToResolveSymbol,
+    onSuccessfullyResolvedSymbol,
+}: {
+    compilerOptions: CompilerOptions;
+    getSymbolOfDeclaration: (node: Declaration) => Symbol;
+    error: (location: Node | undefined, message: DiagnosticMessage, ...args: DiagnosticArguments) => void;
+    globals: SymbolTable;
+    argumentsSymbol: Symbol;
+    requireSymbol: Symbol;
+    lookup: (symbols: SymbolTable, name: __String, meaning: SymbolFlags) => Symbol | undefined;
+    getNodeLinks: (node: Node) => { declarationRequiresScopeChange?: boolean; };

```

Comment:
This does expose NodeLinks in a way that we might not want. We could split this into `setDeclarationRequiresScopeChange` and `getDeclarationRequiresScopeChange` that can be passed in without actually caring how the flag is stored. 

Keeping the cache local would result in different instantiations not being able to share the cache.

Reply:
We could push the logic all the way into a handler you pass in, a `onPossibleVariableResult` which takes the `result` `Symbol`, `location`, and `lastLocation`, and that the checker would just pass in as `useOuterVariableScopeInParameter` - that could keep the links abstraction inside the checker.

Reply:
> We could split this into setDeclarationRequiresScopeChange and getDeclarationRequiresScopeChange that can be passed in without actually caring how the flag is stored.

...that's probably better? I suppose this logic is important for callers to do (just not the particulars of storing the result - or even if the result needs to be stored), so moving it into the checker would result in duplication at all use-sites.